File: internal/collector.go
package internal

import (
	"bytes"
	"context"
	"hysbysu/model"
	"io"
	"log"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/pkg/errors"
	"golang.org/x/sync/errgroup"
)

const (
	el_subjectCard       = ".gallery_grid_item.md-card-content > a"
	el_subjectName       = ".page-header-headings"
	el_topicItem         = ".topics > li .content"
	el_sectionTitle      = ".sectionname"
	el_lectureList       = "ul.section > li"
	el_activityInstance  = ".activityinstance a"
	el_instanceName      = ".instancename"
	el_modtypeResource   = "modtype_resource"
	el_modtypeAssignment = "modtype_assign"
	el_modtypeQuiz       = "modtype_quiz"
	el_modtypeUrl        = "modtype_url"
	el_modtypeForum      = "modtype_forum"
	el_summaryImages     = ".summary img"
	el_lecturerName      = ".summary td strong"
)

type goQueryCollector struct {
	config Config
	client *httpClient
}

var (
	MissingLectureHref  = errors.New("failed to get lecture href property")
	MissingLectureClass = errors.New("failed to get activityinstance class property")
)

func NewCollector(config Config, client *httpClient) *goQueryCollector {
	return &goQueryCollector{config, client}
}

func (c goQueryCollector) CollectSubjectLinks(html io.Reader) ([]string, error) {
	document, err := goquery.NewDocumentFromReader(html)
	if err != nil {
		return nil, errors.Wrap(err, "failed to parse slc document")
	}

	var urls []string
	document.Find(el_subjectCard).Each(func(_ int, a *goquery.Selection) {
		href, exists := a.Attr("href")
		if exists {
			urls = append(urls, href)
		}
	})
	return urls, nil
}

func (c goQueryCollector) collectMeetings(html io.Reader) ([]model.Meeting, error) {
	document, err := goquery.NewDocumentFromReader(html)
	if err != nil {
		return nil, errors.Wrap(err, "failed to collect meetings")
	}

	meetings := make([]model.Meeting, 0)
	subjectName := document.Find(el_subjectName).Text()

	// skip one because the first topicEl is just the detail of the course
	topicEl := document.Find(el_topicItem).Slice(1, goquery.ToEnd)
	for _, topicNode := range topicEl.Nodes {
		selection := goquery.NewDocumentFromNode(topicNode).Selection
		meeting, err := c.extractMeeting(selection)
		if err != nil {
			return nil, errors.Wrap(err, "failed to extract meeting detail")
		}
		meeting.Subject = subjectName
		meetings = append(meetings, meeting)
	}

	return meetings, nil
}

func (c goQueryCollector) extractMeeting(s *goquery.Selection) (meeting model.Meeting, err error) {
	sectionTitle := s.Find(el_sectionTitle).Text()

	lectures := make([]model.Lecture, 0)
	lectureEl := s.Find(el_lectureList)

	for _, lectureNode := range lectureEl.Nodes {
		selection := goquery.NewDocumentFromNode(lectureNode).Selection
		lecture, err := c.extractLecture(selection)
		if err != nil {
			// these errors are not fatal, we'll just ignore them
			if errors.Is(err, MissingLectureHref) || errors.Is(err, MissingLectureClass) {
				continue
			}

			return model.Meeting{}, errors.Wrap(err, "failed to extract meeting")
		}
		lectures = append(lectures, lecture)
	}

	return model.Meeting{
		Subject:  "", // will be filled later
		Title:    sectionTitle,
		Lectures: lectures,
	}, nil
}

func (c goQueryCollector) extractLecture(s *goquery.Selection) (model.Lecture, error) {
	activityInstance := s.Find(el_activityInstance)

	classAttrib, exists := s.Attr("class")
	if !exists {
		return model.Lecture{}, MissingLectureClass
	}

	lectureName := activityInstance.Find(el_instanceName).Text()
	lectureUrl, exists := activityInstance.Attr("href")
	if !exists {
		return model.Lecture{}, MissingLectureHref
	}

	isResource := strings.Contains(classAttrib, el_modtypeResource)
	isAssignment := strings.Contains(classAttrib, el_modtypeAssignment)
	isQuiz := strings.Contains(classAttrib, el_modtypeQuiz)
	isUrl := strings.Contains(classAttrib, el_modtypeUrl)
	isForum := strings.Contains(classAttrib, el_modtypeForum)

	var deadline int64
	var lectureType model.LectureType = model.LectureUnknown

	if isResource {
		lectureType = model.LectureResource
	} else if isAssignment {
		lectureType = model.LectureAssignment
		parsedDeadline, err := c.getAssignmentDeadline(lectureUrl)
		if err != nil {
			return model.Lecture{}, err
		}
		deadline = parsedDeadline
	} else if isQuiz {
		lectureType = model.LectureQuiz
	} else if isUrl {
		lectureType = model.LectureUrl
	} else if isForum {
		lectureType = model.LectureForum
	}

	lecture := model.Lecture{
		Name:     lectureName,
		Url:      lectureUrl,
		Type:     lectureType,
		Deadline: deadline,
	}
	return lecture, nil
}

func (c goQueryCollector) getAssignmentDeadline(lectureUrl string) (int64, error) {
	html, err := c.client.Get(lectureUrl, nil)
	if err != nil {
		return 0, errors.Wrap(err, "failed to fetch lecture detail")
	}
	rawDeadline, err := c.collectDeadlineInfo(html.Body)
	if err != nil {
		return 0, errors.Wrap(err, "failed to extract deadline information")
	}
	parsedDeadline, err := time.ParseInLocation("Monday, 2 January 2006, 15:04 PM", rawDeadline, c.config.TimeZone)
	if err != nil {
		return 0, errors.Wrap(err, "failed to parse deadline")
	}
	return parsedDeadline.In(c.config.TimeZone).UnixMilli(), nil
}

func (c goQueryCollector) collectLecturerInfo(html io.Reader) (model.LecturerInfo, error) {
	document, err := goquery.NewDocumentFromReader(html)
	if err != nil {
		return model.LecturerInfo{}, errors.Wrap(err, "failed to parse ")
	}

	lecturerName := document.Find(el_lecturerName).First().Text()
	// TODO(elianiva): use proper placeholder profile image
	//                 maybe take from the config?
	lecturerImage := document.Find(el_summaryImages).Nodes[1]
	lecturerImageUrl := goquery.NewDocumentFromNode(lecturerImage).AttrOr("src", "https://twirpz.files.wordpress.com/2015/06/twitter-avi-gender-balanced-figure.png")

	// conditionally append the protocol to handle the lecturer image source correctly
	if !strings.HasPrefix(lecturerImageUrl, "https://") {
		lecturerImageUrl = "https:" + lecturerImageUrl
	}

	return model.LecturerInfo{
		Name:     lecturerName,
		ImageUrl: lecturerImageUrl,
	}, nil
}

func (c goQueryCollector) collectDeadlineInfo(html io.Reader) (string, error) {
	document, err := goquery.NewDocumentFromReader(html)
	if err != nil {
		return "", errors.Wrap(err, "cannot create lecture detail document")
	}

	deadlineText := goquery.NewDocumentFromNode(document.Find("tr").Nodes[2]).Find(".lastcol").Text()
	return deadlineText, nil
}

func (c goQueryCollector) CollectSubjects(subjectContent io.Reader) ([]model.Subject, error) {
	links, err := c.CollectSubjectLinks(subjectContent)
	if err != nil {
		return nil, errors.Wrap(err, "failed to collect subject links")
	}

	subjects := make([]model.Subject, 0)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()

	eg, _ := errgroup.WithContext(ctx)
	for _, link := range links {
		courseUrl, err := url.Parse(link)
		courseId := courseUrl.Query().Get("id")
		if err != nil {
			return nil, errors.Wrap(err, "invalid lms url can't be parsed")
		}

		lmsLink := link
		eg.Go(func() error {
			log.Println("collecting data for: " + courseId)
			lmsContent, err := c.client.FetchLmsContent(lmsLink)
			if err != nil {
				return err
			}

			// TODO(elianiva): handle this silly stream->string->stream conversion properly
			//                 we can't consume the reader twice
			lmsContentStr, err := io.ReadAll(lmsContent)
			if err != nil {
				return errors.Wrap(err, "failed to read lms content stream")
			}

			meetings, err := c.collectMeetings(bytes.NewReader(lmsContentStr))
			if err != nil {
				return err
			}

			lectureInfo, err := c.collectLecturerInfo(bytes.NewReader(lmsContentStr))
			if err != nil {
				return err
			}

			subjects = append(subjects, model.Subject{
				Lecturer: lectureInfo,
				CourseId: courseId,
				Meetings: meetings,
			})
			log.Println("done with: " + courseId)

			return nil
		})
	}
	err = eg.Wait()
	if err != nil {
		return nil, errors.Wrap(err, "failed to collect subject")
	}

	return subjects, nil
}

File: worker-api/src/infrastructure/Collector.ts
import * as cheerio from "cheerio";
import type { HttpClient } from "~/application/HttpClient";
import type { ICollector } from "~/application/interfaces/ICollector";
import type { ILogger } from "~/application/interfaces/ILogger";
import {
	LECTURE_TYPE,
	type Lecture,
	type LectureType,
} from "~/business/Lecture";
import type { Lecturer } from "~/business/Lecturer";
import type { Meeting } from "~/business/Meeting";
import type { Subject } from "~/business/Subject";

const SELECTOR = {
	subjectCard: ".gallery_grid_item.md-card-content > a",
	subjectName: ".page-header-headings",
	topicItem: ".topics > li .content",
	sectionTitle: ".sectionname",
	lectureList: "ul.section > li",
	activityInstance: ".activityinstance a",
	instanceName: ".instancename",
	modtypeResource: "modtype_resource",
	modtypeAssignment: "modtype_assign",
	modtypeQuiz: "modtype_quiz",
	modtypeUrl: "modtype_url",
	modtypeForum: "modtype_forum",
	modtypePage: "modtype_page",
	summaryImages: ".summary img",
	lecturerName: ".summary td strong",
	assignmentDeadline:
		".submissionsummarytable > table > tbody > tr:nth-child(3) > td",
} as const;

export class Collector implements ICollector {
	#httpClient: HttpClient;
	#logger: ILogger;

	constructor(httpClient: HttpClient, logger: ILogger) {
		this.#httpClient = httpClient;
		this.#logger = logger;
	}

	#extractLecture(node: cheerio.Cheerio<cheerio.Element>): Lecture | undefined {
		const activityInstance = node.find(SELECTOR.activityInstance);
		const classAttribute = node.attr("class");
		if (classAttribute === undefined) return undefined;

		const name = activityInstance.find(SELECTOR.instanceName).text();
		const url = activityInstance.attr("href");
		if (url === undefined) return undefined;

		let lectureType: LectureType = LECTURE_TYPE.unknown;
		const deadline = undefined;
		switch (true) {
			case classAttribute.includes(SELECTOR.modtypeResource):
				lectureType = LECTURE_TYPE.resource;
				break;
			case classAttribute.includes(SELECTOR.modtypeAssignment):
				lectureType = LECTURE_TYPE.assignment;
				break;
			case classAttribute.includes(SELECTOR.modtypeQuiz):
				lectureType = LECTURE_TYPE.quiz;
				break;
			case classAttribute.includes(SELECTOR.modtypeUrl):
				lectureType = LECTURE_TYPE.url;
				break;
			case classAttribute.includes(SELECTOR.modtypeForum):
				lectureType = LECTURE_TYPE.forum;
				break;
			case classAttribute.includes(SELECTOR.modtypePage):
				lectureType = LECTURE_TYPE.page;
				break;
		}

		return { name, url, type: lectureType, deadline };
	}

	#extractMeeting(element: cheerio.Cheerio<cheerio.Element>): Meeting {
		const sectionTitle = element.find(SELECTOR.sectionTitle).text();
		const $lectures = element.find(SELECTOR.lectureList) ?? [];
		const lectures = $lectures
			.map((_, el) => {
				const htmlText = element.html();
				if (htmlText === null) return undefined;
				const $ = cheerio.load(htmlText, null, false);
				const node = $(el);
				return this.#extractLecture(node);
			})
			.get()
			.filter((lecture) => lecture !== undefined);

		return { subject: "<unknown>", title: sectionTitle, lectures };
	}

	#extractLecturer(html: string): Lecturer {
		const $ = cheerio.load(html);
		let name = $(SELECTOR.lecturerName).first().text();
		let imageUrl =
			$(SELECTOR.summaryImages).children().next().attr("src") ??
			"https://twirpz.files.wordpress.com/2015/06/twitter-avi-gender-balanced-figure.png";

		// conditionally append the protocol to handle the lecturer image source correctly
		if (!imageUrl.startsWith("https://")) {
			imageUrl = `https:${imageUrl}`;
		}

		if (name.length === 0) {
			name = "<unknown>";
		}

		return { name, imageUrl };
	}

	#collectMeetings(rawMeetings: string): {
		subjectName: string;
		meetings: Meeting[];
	} {
		const $ = cheerio.load(rawMeetings);
		const subjectName = $(SELECTOR.subjectName).text();
		// skip one because the first topicEl is just the detail of the course
		return {
			subjectName,
			meetings: $(SELECTOR.topicItem)
				.slice(1)
				.map((_, el) => {
					const node = $(el);
					const meeting = this.#extractMeeting(node);
					meeting.subject = subjectName;
					return meeting;
				})
				.get(),
		};
	}

	#collectAssignmentDeadline(rawLecture: string) {
		const $ = cheerio.load(rawLecture);
		const deadline = $(SELECTOR.assignmentDeadline).text();
		if (deadline.length === 0) return undefined;
		return new Date(deadline);
	}

	public collectSubjectLinks(html: string): string[] {
		const $ = cheerio.load(html);
		return $(SELECTOR.subjectCard)
			.map((_, el) => $(el).attr("href"))
			.get()
			.filter((url) => url !== undefined);
	}

	public async collectSubjects(rawSubjects: string): Promise<Subject[]> {
		const links = this.collectSubjectLinks(rawSubjects).sort();
		const subjects = await Promise.all(
			links.map(async (link) => {
				const url = new URL(link);
				const id = url.searchParams.get("id");
				if (id === null) return undefined;
				this.#logger.info(`collecting content for ${id}`);
				const lmsContent = await this.#httpClient.fetchLmsContent(link);
				const { subjectName, meetings } = this.#collectMeetings(lmsContent);
				const meetingDetailsTask = meetings.map(async (meeting) => {
					const lectureDetailsTask = meeting.lectures.map(async (lecture) => {
						// only need to fetch the details of an assignment
						// we can safely skip if it's not an assignment since they don't have a deadline
						if (lecture.type !== LECTURE_TYPE.assignment) return lecture;
						const assignmentContent =
							await this.#httpClient.fetchAssignmentContent(lecture.url);
						return {
							...lecture,
							deadline: this.#collectAssignmentDeadline(assignmentContent),
						};
					});
					const lectureDetails = await Promise.all(lectureDetailsTask);
					return {
						...meeting,
						lectures: lectureDetails,
					};
				});
				const meetingsWithDetails = await Promise.all(meetingDetailsTask);
				const lecturer = this.#extractLecturer(lmsContent);
				this.#logger.info(`done with ${id}`);
				return {
					title: subjectName,
					lecturer,
					courseId: id,
					meetings: meetingsWithDetails,
				};
			}),
		);
		return subjects.filter(
			(subject): subject is Subject => subject !== undefined,
		);
	}
}
